### PR TITLE
[ownership] Distinguish on any forwarding insts in between the ownership kind that is "forwarded" and the ownership kind of its result to allow for trivial results to be modeled correctly

### PR DESCRIPTION
**Pull request by gottesmm**
Sorry for the bad name of the PR. This PR solves a representation issue that snuck into OSSA as we have made changes: forwarding insts that have a trivial result give the trivial result the specific forwarding ownership in question. This causes the ownership verifier to almost always trip since all uses of the trivial value will expect something with OwnershipKind::None, not something with OwnershipKind::Guaranteed|Owned.

To fix this representation issue this PR does two things:

1. It renames OwnershipForwardingMixin::{get,set}OwnershipKind() -> {get,set}ForwardingOwnershipKind(). This makes it clear when reading the source if one wants to get the ValueOwnershipKind of the result (in which case one uses SILValue::getOwnershipKind) vs the forwarded ownership kind (which also influences/has relationships with the ownership constraint of the operand).
2. It changes SILValue::getOwnershipKind() to return OwnershipKind::None for any forwarding insts with trivial results.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
Makes sense. LGTM.

File: test/SIL/ownership-verifier/use_verifier.sil [hunk 1]
```diff
@@ -1431,3 +1431,30 @@ bb0(%0 : @guaranteed $Builtin.NativeObject):
...
+bb0(%0 : @owned $ThreeDifferingPayloadEnum):
+  // NOTE: It may be surprising that %0 is consumed by this unchecked_enum_data
+  // despite the result of the instruction being trivial... after all... aren't
+  // we leaking something?! If one takes a step back, one will realize that even
+  // though %0 is statically owned, it dynamically is known to be a value with
+  // OwnershipKind::None and thus there is nothing to clean up. So in a certain
+  // sense by converting its non-trivial input type to a trivial output type,
+  // the instruction is stating that it knows that %0 is really dynamically
+  // OwnershipKind::None and thus it is safe to not propagate a value with
+  // non-OwnershipKind::None to be cleaned up to cleanup %0.
```
Comment (meg-gupta) [inline test/SIL/ownership-verifier/use_verifier.sil:1445]: nice explanation, thanks

**Comment by gottesmm** [comment]
@swift-ci test linux platform

**Comment by gottesmm** [comment]
@swift-ci test macOS platform

**Comment by gottesmm** [comment]
@swift-ci test macOS platform

**Comment by gottesmm** [comment]
@swift-ci smoke test macOS platform

**Comment by gottesmm** [comment]
@swift-ci test macOS platform

**Review by atrick — COMMENTED** [review]
@gottesmm lgtm, although `OwnershipForwardingMixin` should be commented to explain that its ownership kind refers to its effect on the operand but is not necessarily the result ownership.

**Pull request merged**